packages/@aws-cdk/aws-cloudformation/test/test.pipeline-actions.ts

```diff
@@ -0,0 +1,144 @@
+import cdk = require('@aws-cdk/cdk');
+import nodeunit = require('nodeunit');
+import util = require('util');
+import cloudformation = require('../lib');
+
+export = nodeunit.testCase({
+  CreateReplaceChangeSet: {
+    works(test: nodeunit.Test) {
+      const stack = new cdk.Stack();
+      const statements = new Array<PolicyStatementJson>();
+      const pipelineRole = {
+        addToPolicy: statement => { statements.push(statement.toJson()); }
+      } as iam.Role;
+      const actions = new Array<cpapi.Action>();
+      const stage = {
+        _attachAction: act => { actions.push(act); },
+        pipelineRole
+      } as cpapi.IStage;
+      const artifact = new cpapi.Artifact(stack as any, 'TestArtifact');
+      const action = new cloudformation.PipelineCreateReplaceChangeSetAction(stack, 'Action', {
+        stage,
+        changeSetName: 'MyChangeSet',
+        stackName: 'MyStack',
+        templatePath: artifact.atPath('path/to/file')
+      });
+
+      test.ok(_grantsPermission(statements, 'iam:PassRole', action.role.roleArn),
+              'The pipelineRole was given permissions to iam:PassRole to the action');
+
+      const stackArn = cdk.ArnUtils.fromComponents({
+        service: 'cloudformation',
+        resource: 'stack',
+        resourceName: 'MyStack/*'
+      });
+      test.ok(_grantsPermission(statements, 'cloudformation:DescribeStacks', stackArn)
+              && _grantsPermission(statements, 'cloudformation:DescribeChangeSet', stackArn),
+              'The pipelineRole was given permissions to describe the stack & it\'s ChangeSets');
+      test.ok(_grantsPermission(statements, 'cloudformation:CreateChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to create the desired ChangeSet');
+      test.ok(_grantsPermission(statements, 'cloudformation:DeleteChangeSet', stackArn, {
+                StringEquals: { 'cloudformation:ChangeSetName': 'MyChangeSet' }
+              }),
+              'The pipelineRole was given permissions to delete the desired ChangeSet');
+
+      test.deepEqual(action.inputArtifacts, [artifact],
+                     'The inputArtifact was correctly registered');
+
+      test.ok(_hasAction(actions, 'AWS', 'CloudFormation', 'Deploy', {
+        ActionMode: 'CHANGE_SET_CREATE_REPLACE',
+        StackName: 'MyStack',
+        ChangeSetName: 'MyChangeSet'
+      }));
+
+      test.done();
+    }
+  },
+  ExecuteChangeSet: {
+    works(test: nodeunit.Test) {
+      const stack = new cdk.Stack();
+      const statements = new Array<PolicyStatementJson>();
+      const pipelineRole = {
+        addToPolicy: statement => { statements.push(statement.toJson()); }
+      } as iam.Role;
+      const actions = new Array<cpapi.Action>();
+      const stage = {
+        _attachAction: act => { actions.push(act); },
+        pipelineRole
+      } as cpapi.IStage;
```

Comment:
Yeah, I know that testing the `codepipeline-api` things suck when you don't have the implementations from `codepipeline` available :(.

Can you at least extract these, so they don't have to be repeated for every test?

I would also accept testing this with an integ test in the `codepipeline` module, and dumping these tests altogether :).

Reply:
I can definitely DRY this up a little. The integration testing from `@aws-cdk/aws-codepipeline` doesn't save me on the minimum test coverage requirements in `@aws-cdk/aws-cloudformation`.